fenzo-core/src/main/java/com/netflix/fenzo/AutoScaler.java

```diff
@@ -205,7 +211,29 @@ private void processScalingNeeds(HostAttributeGroup hostAttributeGroup, Concurre
         int excess = hostAttributeGroup.shortFall>0? 0 : hostAttributeGroup.idleHosts.size() - rule.getMaxIdleHostsToKeep();
         int inactiveIdleCount = hostAttributeGroup.idleInactiveHosts.size();
 
-        if ((inactiveIdleCount > 0 || excess > 0) && shouldScaleDown(now, prevScalingActivity, rule)) {
+        if(inactiveIdleCount > 0 && shouldScaleDownInactive(now, prevScalingActivity, rule)) {
+            ScalingActivity scalingActivity = scalingActivityMap.get(rule.getRuleName());
+            long lastReqstAge = (now - scalingActivity.inactiveScaleDownRequestedAt) / 1000L;
+            if(delayScaleDownBySecs>0L && lastReqstAge > 2 * delayScaleDownBySecs) { // reset the request at time
+                scalingActivity.inactiveScaleDownRequestedAt = now;
+            }
+            else if(delayScaleDownBySecs == 0L || lastReqstAge > delayScaleDownBySecs) {
+                scalingActivity.inactiveScaleDownRequestedAt = 0L;
+                scalingActivity.inactiveScaleDownAt = now;
+                Map<String, String> hostsToTerminate = getInactiveHostsToTerminate(hostAttributeGroup.idleInactiveHosts);
+                StringBuilder sBuilder = new StringBuilder();
+                for (String host : hostsToTerminate.keySet()) {
+                    sBuilder.append(host).append(", ");
+                }
+                logger.info("Scaling down inactive hosts " + rule.getRuleName() + " by "
+                        + hostsToTerminate.size() + " hosts (" + sBuilder.toString() + ")");
+                callback.call(
+                        new ScaleDownAction(rule.getRuleName(), hostsToTerminate.values())
```

Comment:
Do you mean executing callbacks at the end of this method, after all scaling actions are collected?
The alternative - only one callback at a time, may result again in resource access contention between scaling down inactive cluster, and ramping up the new one. This condition happens during task migration between two server groups.